### PR TITLE
Change the "golden path" to use a Docker + Python Virt Env install

### PR DESCRIPTION
**Pull request by hello-binit**
## Description

The "golden path" to install *stretch-ai* has been a Docker + Docker install, where Docker is installed on both the robot and the user's GPU-enabled computer. This path is great for novice users, but not as great for developers looking to build on top of *stretch-ai*. In this PR, I change the "golden path" to use a Docker + Virtual Env install, where Docker is installed on the robot and a Mamba virtual environment is created on the user's computer. 

## Checklist

- \[x\] I have performed a self-review of my code
- \[x\] If it is a core feature, I have added thorough tests
- \[x\] I have added documentation for the changes
- \[x\] I have updated the README file if necessary
- \[x\] I have run on hardware if necessary

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -130,43 +129,6 @@ You can find the prompt used by the LLM at the following location. When running
 
 [./src/stretch/llms/prompts/pickup_prompt.py](./src/stretch/llms/prompts/pickup_prompt.py)
 
-## Installation from Source
```
Comment (hello-cpaxton) [inline README.md:133]: Instead of deleting this please move it to a documetn

File: README.md [hunk 2]
```diff
@@ -24,24 +24,23 @@ Much of the code is licensed under the Apache 2.0 license. See the [LICENSE](LIC
 We recommend the following hardware to run *stretch-ai*. Other GPUs and other versions of Stretch may support some of the capabilities found in this repository, but our development and testing have focused on the following hardware.
 
 - **[Stretch 3](https://hello-robot.com/stretch-3-product) from [Hello Robot](https://hello-robot.com/)**
-  - When *Checking Software*, `stretch_system_check.py` should report that all software passes.
-  - You can find detailed instructions in the [official guide for updating Stretch 3's software](https://docs.hello-robot.com/0.3/software/updating_software/).
```
Comment (hello-cpaxton) [inline README.md:28]: Why is this deleted

File: README.md [hunk 3]
```diff
@@ -24,24 +24,23 @@ Much of the code is licensed under the Apache 2.0 license. See the [LICENSE](LIC
...
+  - When *Checking Hardware*, `stretch_system_check.py` should report that all hardware passes.
 - **Computer with an NVIDIA GPU**
   - The computer should be running Ubuntu 22.04. Later versions might work, but have not been tested.
   - Most of our testing has used a high-end CPU with an NVIDIA GeForce RTX 4090.
 - **Dedicated WiFi access point**
   - Performance depends on high-bandwidth, low-latency wireless communication between the robot and the GPU computer.
   - The official [Stretch WiFi Access Point](https://hello-robot.com/stretch-access-point) provides a tested example.
-
-To use the learning-from-demonstration (LfD) code you'll need the [Stretch Dexterous Teleop Kit](https://hello-robot.com/stretch-dex-teleop-kit).
+- (Optional) [Stretch Dexterous Teleop Kit](https://hello-robot.com/stretch-dex-teleop-kit).
```
Comment (hello-cpaxton) [inline README.md:34]: Why this change

File: README.md [hunk 4]
```diff
@@ -106,7 +105,7 @@ Once you're ready to learn more about **Stretch AI**, you can try out the variet
 
 You can use an LLM to provide free-form text input to the pick and place demo with the `--use_llm` command line argument.
 
-Running the following command will first download an open LLM model. Currently, the default model is [Qwen2.5-3B-Instruct](https://huggingface.co/Qwen/Qwen2.5-3B-Instruct). Running this command downloads ~10GB of data, so you'll need to be patient.
+Running the following command will first download an open LLM model. Currently, the default model is [Qwen2.5-3B-Instruct](https://huggingface.co/Qwen/Qwen2.5-3B-Instruct). Running this command downloads ~10GB of data. Using an ethernet cable instead of Wifi is recommended.
```
Comment (hello-cpaxton) [inline README.md:108]: I don't like this suggestion, it's awkward to plug in ethernet

File: README.md (hunk 1 above)
Comment (hello-cpaxton) [inline README.md:133]: Is this still duplicated somewhere

File: docs/start_with_docker_plus_virtenv.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+## Create a virtual env on your GPU Computer
+
+First, check to see if mamba is installed on your computer:
+
+```bash
+mamba
+```
+
+If you get an "command not found" error, then follow the install instructions for mamba here: https://github.com/conda-forge/miniforge#download
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:135]: Conda forge? Is that allowed?

File: install.sh [hunk 6]
```diff
@@ -89,7 +89,7 @@ echo " - This script will install the following packages:"
 echo "   - pytorch=$PYTORCH_VERSION"
 echo "   - pytorch-cuda=$CUDA_VERSION"
 echo "   - torchvision"
-if [ $INSTALL_TORCH_GEOMETRIC = "true" ]; then
+if [[ $INSTALL_TORCH_GEOMETRIC == "true" ]]; then
```
Comment (hello-cpaxton) [inline install.sh:92]: Why this change? the first version is less bash-specific so seems better

File: docs/start_with_docker_plus_virtenv.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+**Note:** A new env should be created for each new version of *stretch-ai* (each time you `git pull`) using the above instructions.
+
+#### Activate your virtual env
+
+```bash
+mamba activate stretch_ai_<version>
+```
+
+#### Verify NVIDIA Docker
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:201]: Why is this included in the start w.o docker section

File: src/stretch/agent/zmq_client.py [hunk 8]
```diff
@@ -7,7 +7,7 @@
 # Some code may be adapted from other open-source works with their respective licenses. Original
 # license information maybe found below, if so.
 
-# (c) 2024 chris paxton under MIT license
+# (c) 2024 Hello Robot under MIT license
```
Comment (hello-cpaxton) [inline src/stretch/agent/zmq_client.py:10]: is this correct? this is the license from before importing it into this repo

File: docs/start_with_docker_plus_virtenv.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+mamba activate stretch_ai_<version>
+
+# Tell mamba to not use your user sitepackages
+mamba env config vars set PYTHONNOUSERSITE=1
+
+# Install a version of setuptools for which clip works
+pip install setuptools==69.5.1
+
+# Install stretch-ai in "editable" mode so you can develop on it
+pip install -e ./src[dev]
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:171]: do not recommend this! It will cause errors for lots of people!

should be python -m pip install

**Review by hello-cpaxton — CHANGES_REQUESTED** [review]

File: docs/start_with_docker_plus_virtenv.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+# Install system dependencies
+# Pyaudio will fail if these are not installed
+sudo apt-get install libasound-dev portaudio19-dev libportaudio2 libportaudiocpp0 espeak ffmpeg
+
+# Run install script to create a conda environment and install dependencies
+./install.sh
+```
+
+At the prompt to install Detic, say **no**. Take note of the name of the environment. It will be something like `stretch_ai_<version>`.
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:156]: How are you running this if you did not install detic?

File: docs/start_with_docker_plus_virtenv.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+Next, run:
+
+```bash
+# Activate your virtual env
+mamba activate stretch_ai_<version>
+
+# Tell mamba to not use your user sitepackages
+mamba env config vars set PYTHONNOUSERSITE=1
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:165]: Why?

File: README.md (hunk 2 above)
Comment (hello-binit) [reply README.md:28]: Robot-side software is installed in a docker image anyways. We only care that the Checking Hardware section reports all green.

Comment (hello-binit) [reply README.md:28]: See the green part:
Screenshot from 2024-11-14 00-12-39

File: README.md (hunk 3 above)
Comment (hello-binit) [reply README.md:34]: It improves readability. The dex kit is another item on the list.

File: README.md (hunk 4 above)
Comment (hello-binit) [reply README.md:108]: It's awkward, but effective. The robot-side docker was at 10% after 30min, and then finished in 10 min after I plugged in the cable.

Comment (hello-binit) [reply README.md:108]: To be clear, I'm not saying keep it plugged in for the demo. Just the download of the 10GB.

File: README.md (hunk 1 above)
Comment (hello-binit) [reply README.md:133]: Yes here: https://github.com/hello-robot/stretch_ai/pull/176/files#diff-c0f0997fc2d2605b308daa9068fb5ae3a8d7ee877a06df7f2c74f7c384ff0e14

File: docs/start_with_docker_plus_virtenv.md (hunk 5 above)
Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:135]: What do you mean?

File: install.sh (hunk 6 above)
Comment (hello-binit) [reply install.sh:92]: There's a `[: =: unary operator expected` error currently. We can do `[ "$INSTALL_TORCH_GEOMETRIC" = "true" ]` if you prefer non-bash. Fwiw, all modern shells support double bracket.

File: docs/start_with_docker_plus_virtenv.md (hunk 7 above)
Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:201]: Good catch. Removing.

File: docs/start_with_docker_plus_virtenv.md (hunk 9 above)
Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:171]: Don't worry. `pip` is put in your path when you source the virt env.

File: docs/start_with_docker_plus_virtenv.md (hunk 10 above)
Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:156]: It's installed manually in a later step. Detic in the script fails to install for me. See issue https://github.com/hello-robot/stretch_ai/issues/177

File: docs/start_with_docker_plus_virtenv.md (hunk 11 above)
Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:165]: By default virtual envs of the same Python version as your system Python try to use your user site-packages. `PYTHONNOUSERSITE` disables that from happening.

File: docs/start_with_docker_plus_virtenv.md (hunk 9 above)
Comment (hello-cpaxton) [reply docs/start_with_docker_plus_virtenv.md:171]: i have seen many times something go wrong with this

File: docs/start_with_docker_plus_virtenv.md (hunk 10 above)
Comment (hello-cpaxton) [reply docs/start_with_docker_plus_virtenv.md:156]: Why not just fix the script? Or delete it from the script?

Comment (hello-binit) [reply docs/start_with_docker_plus_virtenv.md:156]: I see no harm to leaving it in the script so it can be fixed later.

File: README.md [hunk 12]
```diff
@@ -24,24 +24,23 @@ Much of the code is licensed under the Apache 2.0 license. See the [LICENSE](LIC
...
+  - To use the learning-from-demonstration (LfD) code you'll need the Stretch Dexterous Teleop Kit.
 
 ## Quick-start Guide
 
 Artificial intelligence (AI) for robots often has complex dependencies, including the need for trained models. Consequently, installing *stretch-ai* from source can be challenging.
 
 To help you get started more quickly, we provide two pre-built [Docker](<https://en.wikipedia.org/wiki/Docker_(software)>) images that you can download and use with two shell scripts.
 
-First, you will need to install software on your Stretch robot and another computer with a GPU (*GPU computer*). Use the following link to go to the installation instructions: [Instructions for Installing the Docker Version of Stretch AI](https://github.com/hello-robot/stretch_ai/blob/main/docs/start_with_docker.md)
+First, you will need to install software on your Stretch robot and another computer with a GPU (*GPU computer*). Use the following link to go to the installation instructions: [Instructions for Installing the Docker Version of Stretch AI](https://github.com/hello-robot/stretch_ai/blob/main/docs/start_with_docker_plus_virtenv.md)
```
Comment (hello-cpaxton) [inline README.md:43]: The link title here is not accurate, it's not the docker version of stretch ai

File: docs/start_with_docker_plus_virtenv.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+# Install stretch-ai in "editable" mode so you can develop on it
+pip install -e ./src[dev]
+```
+
+Next, run:
+
+```bash
+# Install detectron2 for perception (required by Detic)
+git submodule update --init --recursive
```
Comment (hello-cpaxton) [inline docs/start_with_docker_plus_virtenv.md:178]: how are you making sure the install script has the same cuda as is installed in mamba? I don't like this recommendation

**Review by hello-cpaxton — CHANGES_REQUESTED** [review]

**Review by hello-cpaxton — APPROVED** [review]
i'll fix errors myself
